core/trino-main/src/main/java/io/trino/sql/planner/RuntimeAdaptivePartitioningRewriter.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.graph.Traverser;
+import io.trino.sql.planner.plan.PlanFragmentId;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.RemoteSourceNode;
+import io.trino.sql.planner.plan.SimplePlanRewriter;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterators.getOnlyElement;
+import static io.trino.sql.planner.SystemPartitioningHandle.FIXED_BROADCAST_DISTRIBUTION;
+import static io.trino.sql.planner.SystemPartitioningHandle.FIXED_HASH_DISTRIBUTION;
+import static io.trino.sql.planner.SystemPartitioningHandle.SCALED_WRITER_HASH_DISTRIBUTION;
+import static io.trino.sql.planner.plan.ExchangeNode.Type.REPLICATE;
+import static io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static java.util.Objects.requireNonNull;
+
+public final class RuntimeAdaptivePartitioningRewriter
+{
+    private RuntimeAdaptivePartitioningRewriter() {}
+
+    public static SubPlan overridePartitionCountRecursively(
+            SubPlan subPlan,
+            int oldPartitionCount,
+            int newPartitionCount,
+            PlanFragmentIdAllocator planFragmentIdAllocator,
+            PlanNodeIdAllocator planNodeIdAllocator,
+            Set<PlanFragmentId> startedFragments)
+    {
+        PlanFragment fragment = subPlan.getFragment();
+        if (startedFragments.contains(fragment.getId())) {
+            // already started, nothing to change for subPlan and its descendants
+            return subPlan;
+        }
+
+        PartitioningScheme outputPartitioningScheme = fragment.getOutputPartitioningScheme();
+        if (outputPartitioningScheme.getPartitioning().getHandle().equals(FIXED_BROADCAST_DISTRIBUTION)) {
```

Comment:
Hmm.

Can it be that we have sth like
```
root
|  
X (output partitioning: FIXED_BROADCAST_DISTRIBUTION)
|
...
|
Y 
|
Z (output partitioning: HASH)
```

if `Z` output is large now then processing of `Y` may fail. 

If such tree shape is not possible can we validate that it actually does not happen with an assertion?

Reply:
Yeah - it feels not very probable - but I also not see a big gain of special casing this. If you want to keep the code please explain in code comment the rationale and why we believe it is ok to do so.

Reply:
Added comment